scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+ * Custom configuration options are determined by checking the metadata.json file.
+ */
+TSGenerator.prototype.generateCustomConfigFromMetadata = function generateCustomConfigFromMetadata(serviceIdentifier) {
+    // some services have additional configuration options that are defined in the metadata.json file
+    // i.e. dualstackAvailable = useDualstack
+    // create reference to custom options
+    var customConfigurations = [];
+    var serviceMetadata = this.metadata[serviceIdentifier];
+    // loop through metadata members
+    for (var memberName in serviceMetadata) {
+        if (!serviceMetadata.hasOwnProperty(memberName)) {
+            continue;
+        }
+        var memberValue = serviceMetadata[memberName];
+        // check configs
+        switch (memberName) {
+            case 'dualstackAvailable':
+                customConfigurations.push(CUSTOM_CONFIG_ENUMS.DUALSTACK);
+                break;
+        }
+    }
+
+    return customConfigurations;
+};
+
+/**
+ * Generates a type or interface based on the shape.
+ */
+TSGenerator.prototype.generateTypingsFromShape = function generateTypingsFromShape(shapeKey, shape, tabCount) {
+    // some shapes shouldn't be generated if they are javascript primitives
+    if (['string', 'boolean', 'number', 'Date', 'Blob'].indexOf(shapeKey) >= 0) {
+        return '';
+    }
+    var self = this;
+    var code = '';
+    tabCount = tabCount || 0;
+    var tabs = this.tabs;
+    var type = shape.type;
+    if (type === 'structure') {
+        code += tabs(tabCount) + 'export interface ' + shapeKey + ' {\n';
+        var members = shape.members;
+        // cycle through members
+        var memberKeys = Object.keys(members);
+        memberKeys.forEach(function(memberKey) {
+            // docs
+            var member = members[memberKey];
+            if (member.documentation) {
+                code += self.generateDocString(member.documentation, tabCount + 1);
+            }
+            var required = self.checkRequired(shape.required || [], memberKey) ? '?' : '';
+            code += tabs(tabCount + 1) + memberKey + required + ': ' + member.shape + ';\n';
+        });
+        code += tabs(tabCount) + '}\n';
+    } else if (type === 'list') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = ' + shape.member.shape + '[];\n';
+    } else if (type === 'map') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = {[key: string]: ' + shape.value.shape + '};\n';
+    } else if (type === 'string' || type === 'character') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = string;\n';
+    } else if (['double', 'long', 'short', 'biginteger', 'bigdecimal', 'integer', 'float'].indexOf(type) >= 0) {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = number;\n';
+    } else if (type === 'timestamp') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = Date;\n';
```

Comment:
If an output field modeled as a timestamp is sourced from the response header rather than the body, then it remains as a string in the output data. For example, in S3, the `LastModified` shape is modeled as a timestamp. In the output of `getObject` and `headObject`, the `LastModified` field is sourced from the header rather than the body, so the value is a string. In `s3.d.ts`, `LastModified` is typed as a Date, which would be incorrect for the output of these two operations. In some other operations such as `listObjects` or `copyObjects`, the `LastModified` field (modeled by the same `LastModified` shape) is sourced from the response body and the value is converted to a Date object.